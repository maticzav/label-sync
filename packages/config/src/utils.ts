export type Dict<T> = { [key: string]: T }

/**
 * Maps entries in an object.
 * @param m
 * @param fn
 */
export function mapEntries<T, V>(m: Dict<T>, fn: (v: T, key: string) => V): Dict<V> {
  const mapped = Object.keys(m).map((key) => {
    return [key, fn(m[key], key)]
  })
  return Object.fromEntries(mapped)
}

/**
 * Maps keys of an object.
 * @param m
 * @param fn
 */
export function mapKeys<T>(m: Dict<T>, fn: (key: string, v: T) => string): Dict<T> {
  const mapped = Object.keys(m).map((key) => {
    return [fn(key, m[key]), m[key]]
  })
  return Object.fromEntries(mapped)
}
