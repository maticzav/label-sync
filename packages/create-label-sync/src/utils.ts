import * as fs from 'fs'
import * as path from 'path'

import { promisify } from 'util'

const mkdir = promisify(fs.mkdir)
const writeFile = promisify(fs.writeFile)

type Dict<T> = { [key: string]: T }

/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(
  fn: (...args: L) => boolean,
): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Loads a tree of utf-8 decoded files at paths.
 *
 * @param path
 */
export function loadTreeFromPath(
  root: string,
  ignore: (string | RegExp)[],
): { [path: string]: string } {
  const files = fs.readdirSync(root, { encoding: 'utf-8' })
  const tree = files
    .filter((file) => !ignore.some((glob) => RegExp(glob).test(file)))
    .flatMap((file) => {
      const rootFilePath = path.resolve(root, file)
      if (fs.lstatSync(rootFilePath).isDirectory()) {
        return Object.entries(
          mapKeys(loadTreeFromPath(rootFilePath, ignore), (key) =>
            unshift(file, key),
          ),
        )
      } else {
        return [[file, fs.readFileSync(rootFilePath, { encoding: 'utf-8' })]]
      }
    })

  return Object.fromEntries(tree)
}

/**
 * Adds a folder to the path.
 * @param pre
 * @param path
 */
function unshift(pre: string, path: string): string {
  return [pre, ...path.split('/').filter(Boolean)].join('/')
}

/**
 * Maps entries in an object.
 * @param m
 * @param fn
 */
export function mapEntries<T, V>(
  m: Dict<T>,
  fn: (v: T, key: string) => V,
): Dict<V> {
  return Object.fromEntries(
    Object.keys(m).map((key) => {
      return [key, fn(m[key], key)]
    }),
  )
}

/**
 * Writes virtual file system representation to the file system.
 *
 * @param vfs
 */
export async function writeTreeToPath(
  root: string,
  tree: { [path: string]: string },
): Promise<void> {
  tree = mapKeys(tree, (file) => path.resolve(root, file))

  const actions = Object.keys(tree).map(async (filePath) => {
    await mkdir(path.dirname(filePath), { recursive: true })
    await writeFile(filePath, tree[filePath])
  })

  try {
    await Promise.all(actions)
  } catch (err) /* istanbul ignore next */ {
    throw err
  }
}

/**
 * Maps keys of an object.
 * @param m
 * @param fn
 */
export function mapKeys<T>(
  m: Dict<T>,
  fn: (key: string, v: T) => string,
): Dict<T> {
  return Object.fromEntries(
    Object.keys(m).map((key) => {
      return [fn(key, m[key]), m[key]]
    }),
  )
}

/**
 * Creates a fallback default value.
 * @param fallback
 * @param value
 */
export function withDefault<T>(fallback: T, value: T | undefined): T {
  if (value) return value
  else return fallback
}
