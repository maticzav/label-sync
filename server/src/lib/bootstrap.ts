import * as handlebars from 'handlebars'
import * as path from 'path'
import * as prettier from 'prettier'

import { mapEntries } from '../data/dict'
import { GHTree } from './github'

/**
 * Populates the template with repositories.
 * @param tree
 */
export function populateTemplate(tree: GHTree, data: { repository: string; repositories: { name: string }[] }): GHTree {
  return mapEntries(tree, (file, name) => {
    /* Personalize file */
    const populatedFile = handlebars.compile(file)(data)

    /* Format it */
    switch (path.extname(name)) {
      case '.ts': {
        return prettier.format(populatedFile, { parser: 'typescript' })
      }
      case '.yml': {
        return prettier.format(populatedFile, { parser: 'yaml' })
      }
      case '.md': {
        return prettier.format(populatedFile, { parser: 'markdown' })
      }
      case '.json': {
        return prettier.format(populatedFile, { parser: 'json' })
      }
      default: {
        return populatedFile
      }
    }
  })
}
