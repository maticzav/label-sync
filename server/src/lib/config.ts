type Config = {
  prod: boolean

  corsOrigins: string[]

  // https://probot.github.io/docs/configuration/
  ghAppId: string
  ghPrivateKey: string
  ghSecret: string

  stripeApiKey: string
  stripeEndpointSecret: string

  databaseUrl: string
  redisUrl: string

  sentryDSN: string
  datadogApiKey: string
}

// Environments

const base = {}

const prod = {
  prod: true,

  corsOrigins: [
    'https://label-sync.com',
    'https://www.label-sync.com',
    'https://webhook.label-sync.com',
  ],

  ghAppId: process.env.GH_APP_ID!,
  ghPrivateKey: process.env.GH_PRIVATE_KEY!,
  ghSecret: process.env.GH_SECRET!,

  stripeApiKey: process.env.STRIPE_API_KEY!,
  stripeEndpointSecret: process.env.STRIPE_ENDPOINT_SECRET!,

  databaseUrl: process.env.DATABASE_URL!,
  redisUrl: process.env.REDIS_URL!,

  sentryDSN: process.env.SENTRY_DSN!,
  datadogApiKey: process.env.DATADOG_API_KEY!,
}

const dev = {
  prod: false,

  corsOrigins: ['http://localhost', 'http://127.0.0.1'],

  ghAppId: '',
  ghPrivateKey: '',
  ghSecret: '',

  stripeApiKey: '',
  stripeEndpointSecret: '',

  databaseUrl: 'postgres://prisma:prisma@localhost:5432/prisma',
  redisUrl: 'redis://localhost:6379',

  sentryDSN: '',
  datadogApiKey: '',
}

const enviroment = process.env.NODE_ENV

/**
 * Configuration credentials for the worker instance.
 */
export const config: Config = Object.assign(base, enviroment === 'production' ? prod : dev)
