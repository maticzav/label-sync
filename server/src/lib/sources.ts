import { InstallationsSource } from '@labelsync/database'
import { TaskQueue } from '@labelsync/queues'
import pino from 'pino'
import Stripe from 'stripe'

/**
 * Datasources shared by many components of the app.
 */
export type Sources = {
  installations: InstallationsSource
  tasks: TaskQueue
  stripe: Stripe
  log: pino.Logger
}
