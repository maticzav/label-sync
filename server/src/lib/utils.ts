/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(fn: (...args: L) => boolean): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Creates a fallback default value.
 * @param fallback
 * @param value
 */
export function withDefault<T>(fallback: T, value: T | undefined | null): T {
  if (value == null) return fallback
  return value
}
