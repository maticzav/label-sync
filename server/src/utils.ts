import fs from 'fs'
import path from 'path'
import { mapKeys } from './data/dict'

/**
 * Negates the wrapped function.
 * @param fn
 */
export function not<T, L extends Array<T>>(
  fn: (...args: L) => boolean,
): (...args: L) => boolean {
  return (...args) => !fn(...args)
}

/**
 * Loads a tree of utf-8 decoded files at paths.
 *
 * @param path
 */
export function loadTreeFromPath(
  root: string,
  ignore: (string | RegExp)[],
): { [path: string]: string } {
  const files = fs.readdirSync(root, { encoding: 'utf-8' })
  const tree = files
    .filter(file => !ignore.some(glob => RegExp(glob).test(file)))
    .flatMap(file => {
      const rootFilePath = path.resolve(root, file)
      if (fs.lstatSync(rootFilePath).isDirectory()) {
        return Object.entries(
          mapKeys(loadTreeFromPath(rootFilePath, ignore), key =>
            unshift(file, key),
          ),
        )
      } else {
        return [[file, fs.readFileSync(rootFilePath, { encoding: 'utf-8' })]]
      }
    })

  return Object.fromEntries(tree)
}

/**
 * Adds a folder to the path.
 * @param pre
 * @param path
 */
function unshift(pre: string, path: string): string {
  return [pre, ...path.split('/').filter(Boolean)].join('/')
}
