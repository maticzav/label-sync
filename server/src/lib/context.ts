import _ from 'lodash'
import { ProbotOctokit } from 'probot'

import { getLSConfigRepoName, LSCConfiguration, LS_CONFIG_PATH, parseConfig } from '@labelsync/config'
import { Installation } from '@labelsync/database'

import { getFile } from './github'
import { Sources } from './sources'
import { DateTime } from 'luxon'
import pino from 'pino'

export type Octokit = InstanceType<typeof ProbotOctokit>

/**
 * Context that all events should include.
 */
interface BaseContext {
  octokit: Octokit
  payload: {
    action?: string | null
    repository: {
      name: string
      default_branch: string
      owner: {
        id: number
        login: string
        type: string
      }
    }
  }
  sources: Sources
}

/**
 * Looks up the installation associated with the account from a given event
 * and checks that the subscription is active.
 */
export async function getAccountInstallation(ctx: BaseContext): Promise<Installation> {
  const account = ctx.payload.repository.owner.login.toLowerCase()

  const installation = await ctx.sources.installations.get({ account })

  /* istanbul ignore if */
  if (installation == null) {
    throw new ExecutionError(`Couldn't find installation for ${account}.`)
  }

  if (DateTime.fromJSDate(installation.periodEndsAt) < DateTime.now()) {
    throw new ExecutionError(`Expired subscription ${installation.account}`)
  }

  return installation
}

/**
 * Returns account configuration related to the provided event.
 */
export async function getAccountConfiguration(ctx: BaseContext): Promise<LSCConfiguration> {
  const installation = await getAccountInstallation(ctx)

  const repo = getLSConfigRepoName(installation.account)
  const ref = `refs/heads/${ctx.payload.repository.default_branch}`

  const rawConfig = await getFile(ctx.octokit, { owner: installation.account, repo, ref }, LS_CONFIG_PATH)

  /* istanbul ignore next */
  if (rawConfig === null) {
    throw new ExecutionError(`No configuration found while loading sources.`)
  }

  const parsedConfig = parseConfig({
    input: rawConfig,
    isPro: installation.plan === 'PAID',
  })

  /* istanbul ignore if */
  if (!parsedConfig.ok) {
    throw new ExecutionError(`Invalid configuration: ${parsedConfig.error}`)
  }

  return parsedConfig.config
}

/**
 * Error that may be safely thrown in the event handler.
 */
export class ExecutionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExecutionError'
  }
}

/**
 * Lets you use a context function in a safe environment.
 */
export async function withContext(ctx: BaseContext & { log: pino.Logger }, fn: () => Promise<void>): Promise<void> {
  try {
    await fn()
  } catch (err) /* istanbul ignore next */ {
    if (process.env.NODE_ENV !== 'production') {
      console.error(err)
      return
    }

    if (err instanceof ExecutionError) {
      ctx.log.debug(
        {
          action: ctx.payload.action,
          repo: ctx.payload.repository.name,
          owner: ctx.payload.repository.owner.login,
        },
        err.message,
      )
      // with context and all
      return
    }

    throw err
  }
}
