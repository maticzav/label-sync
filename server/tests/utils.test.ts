import path from 'path'
import { not, loadTreeFromPath } from '../src/lib/utils'

describe('utils:', () => {
  test('not', () => {
    expect(not(() => true)()).toBeFalsy()
    expect(not(() => false)()).toBeTruthy()
  })

  test('loadTreeFromPath', () => {
    const tree = loadTreeFromPath(path.resolve(__dirname, './__fixtures__/template'), ['ignore'])

    expect(tree).toMatchSnapshot()
  })
})
