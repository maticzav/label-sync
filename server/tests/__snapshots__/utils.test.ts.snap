// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils: loadTreeFromPath 1`] = `
Object {
  "README.md": "# Template Fixture
",
  "src/folder/nested.ts": "export function nested() {
  return true
}
",
  "src/index.ts": "function helloWorld() {
  return 'hi!'
}

helloWorld()
",
  "tsconfig.json": "{
  \\"compilerOptions\\": {
    \\"target\\": \\"es2017\\",
    \\"module\\": \\"commonjs\\",
    \\"lib\\": [\\"dom\\", \\"es2017\\", \\"es2019\\"],
    \\"rootDir\\": \\"src\\",
    \\"outDir\\": \\"dist\\"
  }
}
",
}
`;
